### PR TITLE
added debounce to filter out double click on file

### PR DESCRIPTION
**Pull request by NakulManchanda**
Signed-off-by: Nakul Manchanda <nakul.manchanda@ibm.com>

* added new node click subject to apply debounce operator
* debounce operator helped make double click - behave as single click

### CODE REVIEW

**Comment by DivergentEuropeans** [comment]
Hey @NakulManchanda ,

Thanks for the contribution. I will take a look and review over the weekend, and get back to @1000TurquoisePogs so he can merge. He is away for a few days and will be back on Tuesday.

**Comment by DivergentEuropeans** [comment]
Hey @NakulManchanda 

Can you provide a bit more info on this? I'm not understanding what this is for, or the differences that I see are not immediately apparent, for ex. the double click features that I see such as "drill down" into nodes and close files, and selecting whole words in the text view, these things are all preserved. So what does this do/fix?

**Comment by NakulManchanda** [comment]
Hi,

Try staging branch for existing behavior.

When we double click on file (not folder) it does not render file content, only single click does. This happens because we make two consecutive backend call on double click, and first call when completes it unsubscribes editor service, resulting in blank file in code editor.

Solution is added dbounce time operation, it help ignore multiple clicks and combine them into one call.

Try once, otherwise I can send a video next week by Tuesday.

**Comment by NakulManchanda** [comment]
One more thing, disable all breakpoints when replicating, because it’s a timing issue, it don’t get captured with any breakpoint on click

File: webClient/src/app/editor/project-tree/project-tree.component.ts [hunk 1]
```diff
@@ -130,24 +156,24 @@ export class ProjectTreeComponent implements OnInit {
...
 
     this.editorControl.openDataset.subscribe(dirName => {
       if (dirName != null && dirName !== '') {
         if (dirName[0] == '/') {
           //Note: This temporary hack is used to hide datasets using the original slower Editor structure.
           // Will be removed when Dataset functionality for Explorer gets better.
-            this.fileExplorer.showUss();
-            this.fileExplorer.updateDirectory(dirName);
-            this.showDatasets = false;
+          this.fileExplorer.showUss();
```
Comment (DivergentEuropeans) [inline webClient/src/app/editor/project-tree/project-tree.component.ts]: Generally, we try to shy away from whitespace changes as to not clog up Git any more than it has to be. Personally, I don't have a problem with it but just for future things.

Comment (NakulManchanda) [reply webClient/src/app/editor/project-tree/project-tree.component.ts]: can you share your style guides - like prettier if you are using one, my eslint vscode are formating it automatically, it tried to fix most of them before commit, but then if i modify something again, and hit save, it auto formats, which i like to use

**Comment by NakulManchanda** [comment]
> Functionally works. I don't see any potential issues other than if we decide to add node double click functionality (maybe double click opens properties of file/saves file/anything we come up with), but we don't have that and it can fairly easily be reverted, so this removes nothing and fixes a bug, seems OK.
> 
> The only thing I would like to change is your NgOnDestroy is never called. I checked with both a log and a breakpoint, so either remove that snippet and/or clean up your subscribe in another way. Correct me if I'm wrong with this.

i agree with both points but leave the decision to you guys
1) it is true double click in future can be used for some behavior, but right now it is leading to display blank content, we may be able to find other clever way of resolving the issue, but what i have seen, we have to restrict multiple calls etc, so it can get tricky, in many other double click scenarios, this i found is simple solution for now, and one point of change for all potential double click related scenarios

2) still learning about lifecycle changes in angular, but intention is to automatically clear all subscribers of project tree when it get destroyed (switching to dataset view could be one such scenario).
So my guess is - it will get called implicitly whenever project tree get destroyed. Let me know if you disagree I can remove this, but that in my view is logical place to unsubscribe

File: webClient/src/app/editor/project-tree/project-tree.component.ts (hunk 1 above)
Comment (1000TurquoisePogs) [reply webClient/src/app/editor/project-tree/project-tree.component.ts]: This obviously needs to be expanded: https://zowe.org/code-guidelines-2/
But, generally speaking we just want to minimize the amount of changed lines to maintain git history (the "blame" function).
We tend to use 2 spaces per tab, but I know some other groups may use 4, and it's not as important which you use so much as that you maintain consistency with whatever the current file already uses, and that you don't accidentally cause whitespace changes.

**Comment by DivergentEuropeans** [comment]
> > Functionally works. I don't see any potential issues other than if we decide to add node double click functionality (maybe double click opens properties of file/saves file/anything we come up with), but we don't have that and it can fairly easily be reverted, so this removes nothing and fixes a bug, seems OK.
> > The only thing I would like to change is your NgOnDestroy is never called. I checked with both a log and a breakpoint, so either remove that snippet and/or clean up your subscribe in another way. Correct me if I'm wrong with this.
> 
> i agree with both points but leave the decision to you guys
> 
> 1. it is true double click in future can be used for some behavior, but right now it is leading to display blank content, we may be able to find other clever way of resolving the issue, but what i have seen, we have to restrict multiple calls etc, so it can get tricky, in many other double click scenarios, this i found is simple solution for now, and one point of change for all potential double click related scenarios
> 
> other potential solution is to separate out single and double click, and forward them as separate events down the pipeline, we only need to avoid multiple calls, if you have any suggestions for that, we can re-iterate on this solution
> 
> 1. still learning about lifecycle changes in angular, but intention is to automatically clear all subscribers of project tree when it get destroyed (switching to dataset view could be one such scenario).
>    So my guess is - it will get called implicitly whenever project tree get destroyed. Let me know if you disagree I can remove this, but that in my view is logical place to unsubscribe

No problem, I think we're all learning as we go, myself included ( : Yeah from what I saw, and setting a breakpoint, your unsubscribe code in your NgOnDestroy doesn't get called (if I remember correctly, it's been a few weeks... ) so that's all. Maybe that's supposed to be how that works, but I would expect it to get called when we close that app, although shouldn't happen when we go into a different tab view because the element gets hidden rather than destroyed, in conclusion that's either something really funky going on with Angular and you're doing something incorrectly. There's also a small possibility Angular cleans up subscribes automatically before destroy, which is why your if condition is never satisfied so I don't see it actually hit, but I think it would trigger a check anyway. If you get a chance, feel free to do a little digging (for ex. Do we clean up our subscribes like this normally through the Zowe core apps?) otherwise I'll probably get back to you during the work week some time. You have a quick fix that works, no need to beat around the bush too much

**Comment by NakulManchanda** [comment]
@DivergentEuropeans @1000TurquoisePogs  refreshed this PR by rebasing on current staging, if we agree with premise of this PR to disable double click on nodes, as it was leading to unexpected behavior, please approve this PR, thanks!

**Comment by DivergentEuropeans** [comment]
Seems to have been resolved in staging. Feel free to re-open if it applies!

**Pull request closed without merging**